app/models/mixins/authentication_mixin.rb

```diff
@@ -169,20 +169,15 @@ def update_authentication(data, options = {})
 
       # Update or create
       if cred.nil?
-        if self.kind_of?(ManageIQ::Providers::Openstack::InfraManager) && value[:auth_key]
-          # TODO(lsmola) investigate why build throws an exception, that it needs to be subclass of AuthUseridPassword
-          cred = ManageIQ::Providers::Openstack::InfraManager::AuthKeyPair.new(:name => "#{self.class.name} #{name}", :authtype => type.to_s,
-                                               :resource_id => id, :resource_type => "ExtManagementSystem")
-          authentications << cred
-        elsif value[:auth_key]
-          cred = AuthToken.new(:name => "#{self.class.name} #{name}", :authtype => type.to_s,
-                                               :resource_id => id, :resource_type => "ExtManagementSystem")
-          authentications << cred
-        else
-          cred = authentications.build(:name => "#{self.class.name} #{name}", :authtype => type.to_s,
-                                            :type => "AuthUseridPassword")
-        end
+        cred = authentication_class(value.merge(:authtype => type.to_s)).new(
+          :name          => "#{self.class.name} #{name}",
+          :authtype      => type.to_s,
+          :resource_id   => id,
+          :resource_type => 'ExtManagementSystem'
+        )
+        authentications << cred
```

Comment:
```
-        cred = authentication_class(value.merge(:authtype => type.to_s)).new(
-          :name          => "#{self.class.name} #{name}",
-          :authtype      => type.to_s,
-          :resource_id   => id,
-          :resource_type => 'ExtManagementSystem'
+        auth_klass = authentication_class(value.merge(:authtype => type.to_s))
+        cred = authentications.build(
+          :type     => auth_klass.to_s,
+          :name     => "#{self.class.name} #{name}",
+          :authtype => type.to_s,
+          :resource => self
         )
-        authentications << cred
       end
```

Should fix the one failing host_spec:227 test

Reply:
But adding an auth token fails :(

Okay this method is a mess lets follow-up later to fix it